src/client.rs

```diff
@@ -0,0 +1,371 @@
+    type Response: Serialize + for<'de> Deserialize<'de>;
+    fn as_str<'a>(&'a self) -> &'a str;
+}
+
+#[derive(Debug, serde::Serialize, serde::Deserialize, Deref)]
+pub struct GenericRpcMethod(pub String);
+impl RpcMethod for GenericRpcMethod {
+    type Params = Vec<Value>;
+    type Response = Value;
+    fn as_str<'a>(&'a self) -> &'a str {
+        self.0.as_str()
+    }
+}
+
+#[derive(Debug, serde::Serialize, serde::Deserialize)]
+pub struct RpcRequest<T: RpcMethod> {
+    #[serde(default)]
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub id: Option<Value>,
+    pub method: T,
+    pub params: T::Params,
+}
+
+#[derive(Debug, serde::Serialize, serde::Deserialize)]
+pub struct RpcError {
+    pub code: i64,
+    pub message: String,
+    #[serde(skip)]
+    pub status: Option<StatusCode>,
+}
+impl From<Error> for RpcError {
+    fn from(e: Error) -> Self {
+        RpcError {
+            code: MISC_ERROR_CODE,
+            message: format!("{}", e),
+            status: None,
+        }
+    }
+}
+impl From<serde_json::Error> for RpcError {
+    fn from(e: serde_json::Error) -> Self {
+        RpcError {
+            code: PARSE_ERROR_CODE,
+            message: format!("{}", e),
+            status: None,
+        }
+    }
+}
+impl From<RpcError> for Error {
+    fn from(e: RpcError) -> Self {
+        anyhow!("{}", e.message)
+    }
+}
+
+#[derive(Debug, serde::Serialize, serde::Deserialize)]
+pub struct RpcResponse<T: RpcMethod> {
+    pub id: Option<Value>,
+    pub error: Option<RpcError>,
+    pub result: Option<T::Response>,
+}
+impl From<RpcError> for RpcResponse<GenericRpcMethod> {
+    fn from(e: RpcError) -> Self {
+        RpcResponse {
+            id: None,
+            error: Some(e),
+            result: None,
+        }
+    }
+}
+impl<T: RpcMethod> RpcResponse<T> {
+    pub fn into_result(self) -> Result<T::Response, RpcError> {
+        match self.error {
+            Some(e) => Err(e),
+            None => Ok(self.result).transpose().unwrap_or_else(|| {
+                serde_json::from_value(Value::Null)
+                    .map_err(Error::from)
+                    .map_err(RpcError::from)
+            }),
+        }
+    }
+    pub fn into_response(mut self) -> Result<Response<Body>, Error> {
+        let body = serde_json::to_vec(&self)?;
+        Ok(Response::builder()
+            .status(match self.error.as_mut().and_then(|e| e.status.take()) {
+                Some(s) => s,
+                None if self.error.is_some() => StatusCode::INTERNAL_SERVER_ERROR,
+                None => StatusCode::OK,
+            })
+            .header(CONTENT_LENGTH, body.len())
+            .body(body.into())?)
+    }
+}
+
+#[derive(Debug)]
+pub struct RpcClient {
+    authorization: HeaderValue,
+    uri: Uri,
+    client: HttpClient,
+}
+impl RpcClient {
+    pub fn new(auth: AuthSource, uri: Uri) -> Result<Self, Error> {
+        Ok(RpcClient {
+            authorization: auth.try_load()?,
```

Comment:
This is wrong, `bitcoind` may restart while proxy is running and it'll re-generate a new cookie. Indeed it breaks my deployment. The previous code was intentionally loading the cookie whenever it was needed.

Optimization caching the cookie and reloading&retrying on error is allowed and encouraged.

Reply:
instead of retrying on error, I'm just to invalidating the cache based on the modified timestamp of the cookie file. Hope that works.

Reply:
Hmm, that works although maybe it has smaller performance saving over just reading those few bytes. I guess you're saving at least one syscall (reading is open+read or seek+read; stat is just one syscall).